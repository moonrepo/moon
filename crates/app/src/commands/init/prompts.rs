use super::InitOptions;
use iocraft::prelude::element;
use moon_console::{
    Console,
    ui::{Confirm, Input, Select, SelectOption},
};
use moon_pdk_api::{PromptType, SettingPrompt};
use proto_core::UnresolvedVersionSpec;
use starbase_utils::json::JsonValue;

pub fn fully_qualify_version(version: String) -> Option<String> {
    if version.is_empty() {
        return None;
    }

    let mut parts = version.split('.');

    Some(
        [
            parts.next().unwrap_or("0"),
            parts.next().unwrap_or("0"),
            parts.next().unwrap_or("0"),
        ]
        .join("."),
    )
}

pub async fn render_prompt(
    console: &Console,
    options: &InitOptions,
    prompt: &SettingPrompt,
) -> miette::Result<Option<JsonValue>> {
    match &prompt.ty {
        PromptType::None => Ok(None),
        PromptType::Confirm { default } => {
            let result = if options.yes {
                *default
            } else {
                let mut value = *default;

                console
                    .render_interactive(element! {
                        Confirm(
                            label: &prompt.question,
                            description: prompt.description.clone(),
                            on_confirm: &mut value
                        )
                    })
                    .await?;

                value
            };

            Ok(Some(JsonValue::Bool(result)))
        }
        PromptType::Input { default } => {
            let result = if options.yes {
                default.to_owned()
            } else {
                let mut value = default.to_owned();
                let required = prompt.required;

                console
                    .render_interactive(element! {
                        Input(
                            label: &prompt.question,
                            description: prompt.description.clone(),
                            default_value: default,
                            on_value: &mut value,
                            validate: move |input: String| {
                                if input.is_empty() && required {
                                    Some("Please provide a value".into())
                                } else {
                                    None
                                }
                            }
                        )
                    })
                    .await?;

                value
            };

            Ok(Some(JsonValue::String(result)))
        }
        PromptType::Select {
            default_index,
            options: items,
        } => {
            let index = if options.yes {
                *default_index
            } else {
                let mut index = *default_index;

                console
                    .render_interactive(element! {
                        Select(
                            label: &prompt.question,
                            description: prompt.description.clone(),
                            default_index: *default_index,
                            on_index: &mut index,
                            options: items.iter().map(|i| SelectOption::new(display_json_value(i))).collect::<Vec<_>>()
                        )
                    })
                    .await?;

                index
            };

            Ok(Some(items[index].clone()))
        }
    }
}

pub async fn render_version_prompt(
    console: &Console,
    options: &InitOptions,
    tool: &str,
    op: impl FnOnce() -> miette::Result<Option<UnresolvedVersionSpec>>,
) -> miette::Result<Option<UnresolvedVersionSpec>> {
    let default_version = op()?;

    if options.yes || options.minimal {
        return Ok(default_version);
    }

    let mut confirmed = false;
    let mut value = String::new();

    console
        .render_interactive(element! {
            Confirm(
                label: if let Some(version) = &default_version {
                    format!(
                        "Manage {tool} {version} through <shell>moon</shell>? <muted>(recommended)</muted>"
                    )
                } else {
                    format!(
                        "Manage {tool} through <shell>moon</shell>? <muted>(recommended)</muted>"
                    )
                },
                description: "Will download and install on-demand.".to_string(),
                on_confirm: &mut confirmed,
            )
        })
        .await?;

    if confirmed {
        console
            .render_interactive(element! {
                Input(
                    label: format!("{tool} version?"),
                    default_value: default_version.map(|v| v.to_string()).unwrap_or_default(),
                    on_value: &mut value,
                    validate: move |input: String| {
                        if input.trim().is_empty() {
                            Some("Please provide a version".into())
                        } else if let Err(error) = UnresolvedVersionSpec::parse(&input) {
                            Some(error.to_string())
                        } else {
                            None
                        }
                    }
                )
            })
            .await?;
    }

    Ok(if value.is_empty() {
        None
    } else {
        UnresolvedVersionSpec::parse(value).ok()
    })
}

pub fn display_json_value(value: &JsonValue) -> String {
    match value {
        // Remove quotes
        JsonValue::String(string) => string.to_owned(),
        other => other.to_string(),
    }
}
