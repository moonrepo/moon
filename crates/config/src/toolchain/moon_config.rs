use crate::config_struct;
use schematic::{Config, validate};

config_struct!(
    /// Configures how and where updates will be received.
    #[derive(Config)]
    pub struct MoonConfig {
        /// A secure URL to lookup the latest version.
        #[setting(validate = validate::url_secure, default = "https://launch.moonrepo.app/versions/cli/current")]
        pub manifest_url: String,

        /// A secure URL for downloading the moon binary.
        #[setting(validate = validate::url_secure, default = "https://github.com/moonrepo/moon/releases/latest/download")]
        pub download_url: String,
    }
);
