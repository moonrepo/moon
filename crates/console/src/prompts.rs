use crate::console::Console;
use inquire::error::InquireResult;
use inquire::ui::{
    Attributes, Color as UiColor, ErrorMessageRenderConfig, RenderConfig, StyleSheet, Styled,
};
use moon_common::color::{Color, no_color};
use std::fmt::Display;

pub use inquire::*;

// Add implementations specific to prompts so that they work together
// without much issue. The biggest problem is that problems write
// directly to stderr, while our logs are buffered. To work around this
// we flush stderr before prompting.
//
// It also allows us to apply styles to all prompts from here instead
// of each callsite, which is super nice!
impl Console {
    fn handle_prompt<T>(&self, result: InquireResult<T>) -> miette::Result<T> {
        result.map_err(|error| miette::miette!(code = "console::prompt", "{}", error.to_string()))
    }

    pub fn confirm(&self, prompt: Confirm) -> miette::Result<bool> {
        self.err.flush()?;
        self.handle_prompt(prompt.with_render_config(*self.theme()).prompt())
    }

    pub fn prompt_custom<T: Clone>(&self, prompt: CustomType<T>) -> miette::Result<T> {
        self.err.flush()?;
        self.handle_prompt(prompt.with_render_config(*self.theme()).prompt())
    }

    pub fn prompt_multiselect<T: Display>(&self, prompt: MultiSelect<T>) -> miette::Result<Vec<T>> {
        self.err.flush()?;
        self.handle_prompt(prompt.with_render_config(*self.theme()).prompt())
    }

    pub fn prompt_select<T: Display>(&self, prompt: Select<T>) -> miette::Result<T> {
        self.err.flush()?;
        self.handle_prompt(prompt.with_render_config(*self.theme()).prompt())
    }

    pub fn prompt_select_skippable<T: Display>(
        &self,
        prompt: Select<T>,
    ) -> miette::Result<Option<T>> {
        self.err.flush()?;
        self.handle_prompt(prompt.with_render_config(*self.theme()).prompt_skippable())
    }

    pub fn prompt_text(&self, prompt: Text) -> miette::Result<String> {
        self.err.flush()?;
        self.handle_prompt(prompt.with_render_config(*self.theme()).prompt())
    }
}

fn rgb(color: Color) -> UiColor {
    UiColor::AnsiValue(color as u8)
}

pub fn create_theme() -> RenderConfig<'static> {
    if no_color() {
        return RenderConfig::empty()
            .with_answer(StyleSheet::new().with_attr(Attributes::BOLD))
            .with_prompt_prefix(Styled::new("›"))
            .with_answered_prompt_prefix(Styled::new("✔"))
            .with_scroll_up_prefix(Styled::new("▴"))
            .with_scroll_down_prefix(Styled::new("▾"))
            .with_highlighted_option_prefix(Styled::new("›"))
            .with_canceled_prompt_indicator(Styled::new("(skipped)"))
            .with_selected_checkbox(Styled::new("◉"))
            .with_unselected_checkbox(Styled::new("◯"))
            .with_error_message(ErrorMessageRenderConfig::empty().with_prefix(Styled::new("✘")));
    }

    RenderConfig::empty()
        // Inputs
        .with_default_value(StyleSheet::new().with_fg(rgb(Color::Pink)))
        .with_answer(
            StyleSheet::new()
                .with_fg(rgb(Color::Purple))
                .with_attr(Attributes::BOLD),
        )
        // Prefixes
        .with_prompt_prefix(Styled::new("›").with_fg(rgb(Color::Blue)))
        .with_answered_prompt_prefix(Styled::new("✔").with_fg(rgb(Color::Green)))
        .with_scroll_up_prefix(Styled::new("▴").with_fg(rgb(Color::GrayLight)))
        .with_scroll_down_prefix(Styled::new("▾").with_fg(rgb(Color::GrayLight)))
        .with_highlighted_option_prefix(Styled::new("›").with_fg(rgb(Color::Teal)))
        // States
        .with_help_message(StyleSheet::new().with_fg(rgb(Color::Purple)))
        .with_error_message(
            ErrorMessageRenderConfig::empty()
                .with_prefix(Styled::new("✘").with_fg(rgb(Color::Red)))
                .with_message(StyleSheet::new().with_fg(rgb(Color::Red))),
        )
        .with_canceled_prompt_indicator(Styled::new("(skipped)").with_fg(rgb(Color::Gray)))
        // Selects
        .with_selected_option(Some(StyleSheet::new().with_fg(rgb(Color::Teal))))
        .with_selected_checkbox(Styled::new("◉").with_fg(rgb(Color::Teal)))
        .with_unselected_checkbox(Styled::new("◯").with_fg(rgb(Color::GrayLight)))
}
