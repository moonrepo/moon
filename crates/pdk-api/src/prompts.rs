use serde_json::Value as JsonValue;
use warpgate_api::{api_enum, api_struct};

api_enum!(
    /// The type of prompt to render to receive an answer.
    #[derive(Default)]
    #[serde(tag = "type", rename_all = "kebab-case")]
    pub enum PromptType {
        #[default]
        #[doc(hidden)]
        None,

        /// A confirmation prompt.
        Confirm { default: bool },

        /// A text input field.
        Input { default: String },

        /// A select field with options.
        Select {
            default_index: usize,
            options: Vec<JsonValue>,
        },
    }
);

// api_struct!(
//     pub struct Prompt {
//         pub question: String,
//         pub required: bool,
//         pub ty: PromptType,
//     }
// );

api_struct!(
    /// Represents a prompt (question) for a configuration setting.
    #[serde(default)]
    pub struct SettingPrompt {
        /// A condition to evaluate on whether to render this prompt.
        pub condition: Option<SettingCondition>,

        /// Description of what the setting will do.
        pub description: Option<String>,

        /// Will be rendered in the minimal initialization flow.
        pub minimal: bool,

        /// Nested prompts to render if the answer is truthy.
        pub prompts: Vec<SettingPrompt>,

        /// The question to prompt the user.
        pub question: String,

        /// Whether this prompt is required or optional.
        pub required: bool,

        /// Name of the setting to inject. Supports dot notation.
        pub setting: String,

        /// Skip injecting this setting if the answer is falsy.
        pub skip_if_falsy: bool,

        /// Type of prompt to render.
        pub ty: PromptType,
    }
);

impl SettingPrompt {
    /// Create a new minimal setting.
    pub fn new(setting: impl AsRef<str>, question: impl AsRef<str>, ty: PromptType) -> Self {
        Self {
            minimal: true,
            question: question.as_ref().into(),
            required: true,
            setting: setting.as_ref().into(),
            ty,
            ..Default::default()
        }
    }

    /// Create a new full (non-minimal) setting.
    pub fn new_full(setting: impl AsRef<str>, question: impl AsRef<str>, ty: PromptType) -> Self {
        let mut prompt = Self::new(setting, question, ty);
        prompt.minimal = false;
        prompt
    }
}

api_enum!(
    /// A type of condition to evaluate against a setting value.
    #[derive(Default)]
    #[serde(tag = "op", content = "match", rename_all = "kebab-case")]
    pub enum ConditionType {
        BoolEquals(bool),
        #[default]
        Exists,
        FloatEquals(f64),
        IntEquals(i64),
        NotExists,
        StringContains(String),
        StringEquals(String),
    }
);

api_struct!(
    /// Represents a condition against another setting.
    pub struct SettingCondition {
        pub op: ConditionType,
        pub setting: String,
    }
);
