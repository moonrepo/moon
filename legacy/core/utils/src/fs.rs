use crate::time::{is_stale, to_millis};
use moon_common::consts::CONFIG_DIRNAME;
use starbase_utils::fs;
use std::path::{Path, PathBuf};
use std::time::Duration;

pub mod temp {
    use super::*;

    pub fn get_dir() -> PathBuf {
        crate::get_workspace_root()
            .join(CONFIG_DIRNAME)
            .join("cache")
            .join("temp")
    }

    pub fn get_file(source: &str, ext: &str) -> PathBuf {
        get_dir().join(format!("{}.{}", crate::hash(source), ext))
    }

    pub fn read<P: AsRef<Path>>(path: P) -> miette::Result<Option<String>> {
        let file = path.as_ref();

        if !file.exists() {
            return Ok(None);
        }

        if let Ok(metadata) = file.metadata()
            && let Ok(filetime) = metadata.created()
        {
            // Temp files only last for 4 hours (half a workday)
            if is_stale(to_millis(filetime), Duration::from_secs(60 * 60 * 4)) {
                fs::remove_file(file)?;

                return Ok(None);
            }
        }

        Ok(Some(fs::read_file(file)?))
    }

    pub fn write<P: AsRef<Path>, D: AsRef<str>>(path: P, data: D) -> miette::Result<()> {
        let path = path.as_ref();

        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::write_file(path, data.as_ref())?;

        Ok(())
    }
}
