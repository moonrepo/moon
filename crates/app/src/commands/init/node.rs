use super::InitOptions;
use super::prompts::*;
use iocraft::prelude::element;
use miette::IntoDiagnostic;
use moon_config::load_toolchain_node_config_template;
use moon_console::{
    Console,
    ui::{Container, Entry, Section, Style, StyledText},
};
use moon_lang::{is_using_dependency_manager, is_using_version_manager};
use moon_node_lang::PackageJsonCache;
use moon_pdk_api::{PromptType, SettingPrompt};
use proto_core::UnresolvedVersionSpec;
use starbase_utils::fs;
use starbase_utils::json::JsonValue;
use tera::{Context, Tera};
use tracing::instrument;

pub fn render_template(context: Context) -> miette::Result<String> {
    Tera::one_off(load_toolchain_node_config_template(), &context, false).into_diagnostic()
}

fn detect_node_version(options: &InitOptions) -> miette::Result<Option<UnresolvedVersionSpec>> {
    Ok(if is_using_version_manager(&options.dir, ".nvmrc") {
        UnresolvedVersionSpec::parse(fs::read_file(options.dir.join(".nvmrc"))?).ok()
    } else if is_using_version_manager(&options.dir, ".node-version") {
        UnresolvedVersionSpec::parse(fs::read_file(options.dir.join(".node-version"))?).ok()
    } else {
        None
    })
}

fn detect_node_version_manager(options: &InitOptions) -> miette::Result<String> {
    Ok(if is_using_version_manager(&options.dir, ".nvmrc") {
        "nvm".to_owned()
    } else if is_using_version_manager(&options.dir, ".node-version") {
        "nodenv".to_owned()
    } else {
        String::new()
    })
}

async fn detect_package_manager(
    console: &Console,
    options: &InitOptions,
) -> miette::Result<(String, Option<UnresolvedVersionSpec>)> {
    let mut pm_type = String::new();
    let mut pm_version = String::new();

    // Extract value from `packageManager` field
    if let Ok(Some(pkg)) = PackageJsonCache::read(&options.dir) {
        if let Some(pm) = pkg.data.package_manager {
            if pm.contains('@') {
                let mut parts = pm.split('@');

                pm_type = parts.next().unwrap_or_default().to_owned();
                pm_version = parts.next().unwrap_or_default().to_owned();

                // Remove corepack hash
                if let Some(index) = pm_version.find('+') {
                    pm_version = pm_version[0..index].to_owned();
                }
            } else {
                pm_type = pm;
            }
        }
    }

    // If no value, detect based on files
    if pm_type.is_empty() {
        if is_using_dependency_manager(&options.dir, "yarn.lock") {
            pm_type = "yarn".to_owned();
        } else if is_using_dependency_manager(&options.dir, "pnpm-lock.yaml") {
            pm_type = "pnpm".to_owned();
        } else if is_using_dependency_manager(&options.dir, "bun.lockb") {
            pm_type = "bun".to_owned();
        } else if is_using_dependency_manager(&options.dir, "package-lock.json") {
            pm_type = "npm".to_owned();
        }
    }

    // If no value again, ask for explicit input
    if pm_type.is_empty() {
        let pm = render_prompt(
            console,
            options,
            &SettingPrompt::new(
                "packageManager",
                "Package manager?",
                PromptType::Select {
                    default_index: 0,
                    options: vec!["npm".into(), "pnpm".into(), "yarn".into(), "bun".into()],
                },
            ),
        )
        .await?;

        if let Some(JsonValue::String(inner)) = pm {
            pm_type = inner;
        }
    }

    let pm_version = render_version_prompt(console, options, &pm_type, || {
        if pm_version.is_empty() {
            Ok(None)
        } else {
            Ok(UnresolvedVersionSpec::parse(&pm_version).ok())
        }
    })
    .await?;

    Ok((pm_type, pm_version))
}

#[instrument(skip_all)]
pub async fn init_node(console: &Console, options: &InitOptions) -> miette::Result<String> {
    if !options.yes {
        console.render(element! {
            Container {
                Section(title: "Node") {
                    Entry(
                        name: "Toolchain",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/concepts/toolchain",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                    Entry(
                        name: "Handbook",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/guides/javascript/node-handbook",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                    Entry(
                        name: "Config",
                        value: element! {
                            StyledText(
                                content: "https://moonrepo.dev/docs/config/toolchain#node",
                                style: Style::Url
                            )
                        }.into_any()
                    )
                }
            }
        })?;
    }

    let node_version =
        render_version_prompt(console, options, "Node", || detect_node_version(options)).await?;
    let node_version_manager = detect_node_version_manager(options)?;
    let package_manager = detect_package_manager(console, options).await?;

    let infer_tasks = render_prompt(
        console,
        options,
        &SettingPrompt::new(
            "inferTasks",
            "Infer <file>package.json</file> scripts as moon tasks? <muted>(not recommended)</muted>",
            PromptType::Confirm { default: false },
        ),
    )
    .await?;

    let sync_dependencies = render_prompt(
        console,
        options,
        &SettingPrompt::new(
            "syncDependencies",
            "Sync project relationships as <file>package.json</file> <property>dependencies</property>?",
            PromptType::Confirm { default: true },
        ),
    )
    .await?;

    let dedupe_lockfile = render_prompt(
        console,
        options,
        &SettingPrompt::new(
            "dedupeLockfile",
            "Automatically dedupe lockfile when changed?",
            PromptType::Confirm { default: true },
        ),
    )
    .await?;

    let mut context = Context::new();
    if let Some(node_version) = node_version {
        context.insert("node_version", &node_version);
    }
    context.insert("node_version_manager", &node_version_manager);
    context.insert("package_manager", &package_manager.0);
    if let Some(package_manager_version) = package_manager.1 {
        context.insert("package_manager_version", &package_manager_version);
    }
    context.insert("infer_tasks", &infer_tasks);
    context.insert("sync_dependencies", &sync_dependencies);
    context.insert("dedupe_lockfile", &dedupe_lockfile);
    context.insert("minimal", &options.minimal);

    render_template(context)
}

#[cfg(test)]
mod tests {
    use super::*;
    use starbase_sandbox::assert_snapshot;

    fn create_context() -> Context {
        let mut context = Context::new();
        context.insert("node_version", &"16.0.0");
        context.insert("node_version_manager", &"");
        context.insert("package_manager", &"npm");
        context.insert("package_manager_version", &"8.0.0");
        context.insert("infer_tasks", &false);
        context.insert("dedupe_lockfile", &false);
        context.insert("sync_dependencies", &true);
        context.insert("minimal", &false);
        context
    }

    #[test]
    fn renders_default() {
        assert_snapshot!(render_template(create_context()).unwrap());
    }

    #[test]
    fn renders_minimal() {
        let mut context = create_context();
        context.insert("minimal", &true);

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_nvm() {
        let mut context = create_context();
        context.insert("node_version_manager", &"nvm");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_nodenv() {
        let mut context = create_context();
        context.insert("node_version_manager", &"nodenv");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_npm() {
        let mut context = create_context();
        context.insert("package_manager", &"npm");
        context.insert("package_manager_version", &"9.0.0");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_pnpm() {
        let mut context = create_context();
        context.insert("package_manager", &"pnpm");
        context.insert("package_manager_version", &"7.14.0");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_yarn() {
        let mut context = create_context();
        context.insert("package_manager", &"yarn");
        context.insert("package_manager_version", &"3.2.0");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_bun() {
        let mut context = create_context();
        context.insert("package_manager", &"bun");
        context.insert("package_manager_version", &"1.0.0");

        assert_snapshot!(render_template(context).unwrap());
    }

    #[test]
    fn renders_tasks() {
        let mut context = create_context();
        context.insert("infer_tasks", &true);

        assert_snapshot!(render_template(context).unwrap());
    }
}
