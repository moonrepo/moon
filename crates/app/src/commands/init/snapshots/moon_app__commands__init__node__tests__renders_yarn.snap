---
source: crates/app/src/commands/init/node.rs
expression: render_template(context).unwrap()
---
node:
  version: '16.0.0'
  packageManager: 'yarn'
  yarn:
    version: '3.2.0'
  addEnginesConstraint: true
  dedupeOnLockfileChange: false
  # dependencyVersionFormat: 'workspace'
  inferTasksFromScripts: false
  # rootPackageOnly: true
  syncProjectWorkspaceDependencies: true
  # syncVersionManagerConfig: 'nvm'
